Add BIT_PACKED level encoder/decoder

sadikovi opened the pull request:
This PR adds support for BIT_PACKED level encoder/decoder alongside RLE level encoder/decoder. Bit packed level decoder/encoder is treated as dev/null, and it is used mainly as placeholder for repetition level 0 -> [parquet-mr dev/null encoder](https://github.com/apache/parquet-mr/blob/master/parquet-column/src/main/java/org/apache/parquet/column/values/bitpacking/DevNullValuesWriter.java). I may be wrong with dev/null observation, but found this behaviour on several files - then repetition level > 0 - RLE, otherwise BIT_PACKED.

This implementation is working (not dummy like in parquet-mr), and allows to write and read levels with bit packing, similar to parquet-cpp.

**Implementation notes**:
- `LevelDecoder` and `LevelEncoder` are now enums for RLE and BIT_PACKED
- `set_data()` method now takes `num_buffered_values` to calculate max length to decode for bit packed decoder, also value is used to keep track of values left
- `set_data_range()` method now takes `num_buffered_values` to keep track of total values left
- `get()` method is implemented as simple for loop, I added TODO to implement `get_batch()`

I added unit-tests for level encoding/decoding. I could not test bit packed level encoding/decoding when max level is greater than 0 on actual files (they are all 0 when it is bit packed).

sadikovi commented:
@sunchao I would appreciate if you could have a look at this pull request. Thanks!

sadikovi commented:
@sunchao could you also have a look at this work? I made these changes since some of the files end up with bit packed level encoding. Thanks!

sunchao commented:
@sadikovi sure. Will take a look soon.

sadikovi commented on src/encodings/levels.rs:
```diff
@@ -15,44 +15,64 @@
+/// A encoder for definition/repetition levels.
+/// Currently only supports RLE and BIT_PACKED (dev/null) encoding.
+pub enum LevelEncoder {
+  LevelEncoderRLE { bit_width: u8, rle_encoder: RleEncoder },
```
Yes, sure. I will rename both RLE and Bit packed.

sadikovi commented:
Should I add more tests? I feel like there are some cases that I have not covered.

sunchao commented:
> Should I add more tests? I feel like there are some cases that I have not covered.

The current tests look OK for me. The only things I can think of are:
1. in `LevelEncoder.put()`, when `rle_encoder.put()` or `bit_packed_encoder.put_value()` calls return false, therefore `num_decoded` != `buffer.len()`. This doesn't seem to be covered right now.
2. similarly, in `LevelDecoder.get()`, when `buffer.len()` > number of values in the decoder.

BTW: some documents may be needed for the public functions of LevelDecoder/LevelEncoder. But I think we can address that in a separate PR.

Let me know if you have other thoughts :)

sadikovi commented:
Yes, you are right - I will add test cases. Also agree that we should document functions - will do that.

sadikovi commented on src/encodings/levels.rs:
This is similar to RLE, but in RLE the error is thrown from `rle_encoding.rs`.

sunchao replied:
Makes sense. I think `rle_encoder` returns `Err` when the internal buffer is full, which is confusing. I was not consistent in the return value: when the buffer is full, sometimes we return false, while sometimes we return `Err`. In future maybe we should add a new error type such as `BufferOverflow`. But this doesn't have to be addressed in this PR.

sadikovi commented on src/encodings/levels.rs:
```diff
@@ -64,103 +97,274 @@ impl LevelEncoder {
+      },
+      LevelDecoder::BIT_PACKED { bit_width, ref mut num_values, ref mut bit_packed_decoder } => {
+        // check if set data is called
+        assert!(num_values.is_some());
```
Similar to RLE, we check if data is available to read.

sadikovi commented on src/encodings/levels.rs:
```diff
@@ -64,103 +97,274 @@ impl LevelEncoder {
+          }
+        }
+        // update current values
+        *num_values = num_values.map(|x| x - i);
```
Used to keep track of current number of values.

sadikovi commented on src/util/bit_util.rs:
```diff
@@ -207,7 +207,7 @@ impl BitWriter {
   pub fn skip(&mut self, num_bytes: usize) -> Result<usize> {
     self.flush();
-    assert!(self.byte_offset < self.max_bytes);
+    assert!(self.byte_offset <= self.max_bytes);
```
This seems to be the correct check, arrow bit writer has assertion on `less than or equal`.

sunchao replied:
Yes this looks right.

sadikovi commented:
@sunchao I added tests and documentation. Would appreciate if you could review them when you have time. 

I left some inline comments on pull request, and overall changes are below:

- Both RLE and BIT_PACKED level decoders have `num_values` field to keep track of total values left in decoder. I found that RLE and bit packed can return more values than there are actually (those additional values are zeros) - added test for it
- Updated several assertion checks
- Made `set_data` for data page v1 and `set_data_range` for data page v2 to take `num_buffered_values` as total number of values that decoder has
- This allows me to add assertion on whether or not `set_data` was called before getting values (initially it was only supported by RLE)

Honestly, these changes become larger and larger, and I am not very confident in my code changes. Could you help me to double check any performance implications or better handling of situations when we need to check if set is called before get, and values increment, etc.? Thanks!

_I will run manual tests on several files to make sure I can still read them with my changes:)._

sunchao commented on src/encodings/levels.rs:
```diff
@@ -64,120 +97,352 @@ impl LevelEncoder {
+        // when extracting values from bit reader, it might return more values than left because of
+        // padding to a full byte, we use num_values to track precise number of values.
+        // TODO: Use get_batch() for bit packed decoder
+        let mut i = 0;
```
nit: maybe rename `i` to something more meaningful like `values_read`?

sunchao commented on src/encodings/levels.rs:
```diff
@@ -64,120 +97,352 @@ impl LevelEncoder {
+/// A decoder for definition/repetition levels.
+/// Currently only supports RLE and BIT_PACKED (dev/null) encoding.
+pub enum LevelDecoder {
+  RLE { bit_width: u8, num_values: Option<usize>, rle_decoder: RleDecoder },
```
Seems there are some duplication. Maybe we can create an enum just for the internal decoder? i.e.:
```
enum InternalDecoder {
  RLE { decoder: RleDecoder },
  BIT_PACKED { decoder: BitRaeder }
}
```

This also applies to `LevelEncoder`.

sunchao commented on src/encodings/levels.rs:
```diff
@@ -64,120 +97,352 @@ impl LevelEncoder {
+    // insert a large number of values, so we run out of space
+    for _ in 0..100 {
+      match encoder.put(&levels) {
+        Ok(_) => found_err = false,
```
I guess OK branch can just be a no-op.

sunchao commented on src/encodings/levels.rs:
```diff
@@ -64,120 +97,352 @@ impl LevelEncoder {
+    // perform incremental read until all bytes are read
+    while safe_stop > 0 {
+      safe_stop -= 1;
+      let mut tmp = vec![0; 1];
```
not sure if `tmp` is necessary - we can just use `&mut buffer[total_decoded..total_decoded+1]`

sadikovi commented:
@sunchao thanks for the review! I will address comments.

sadikovi commented:
@sunchao I updated the code, looks better now. I found that it fails when reading one of the examples I have, will try fixing it. Meanwhile, could you have a look, when you have time? Thanks!

sadikovi commented:
@sunchao I updated function documentation - should be okay now. Also issue was related to repetition levels which is fixed in master, after rebase all my example files worked correctly.

sadikovi commented:
@sunchao Should we merge this PR?:)

sunchao commented:
Merged. Thanks @sadikovi for the awesome work!

sadikovi commented:
@sunchao thank you for reviews and merge!
